Sam/world boundaries

samjgorman opened the pull request:
### Description of Changes
Introduce world boundaries with addition of boundaries.tsx and modification of floor.tsx. 
Refactor shapes library for clarity, introduce new props to make shapes translucent.
Introduced a changelog for us to maintain over time.

![image](https://user-images.githubusercontent.com/31233283/163704133-e2eb3817-4e23-4d5f-ba75-3697601e91b9.png)


### Link to Notion ticket
https://www.notion.so/irisapp/Add-a-world-boundary-and-define-behavior-at-end-of-the-world-4971961ebda04cc397830963ac458803

### Testing Procedure
Tested each prop in boundaries & shapes lib locally in combination with other props to ensure expected behavior. Modified world constants for height and width to ensure that boundaries resize correctly.

samjgorman commented:
@kanyesthaker  @shalins take a look!

kanyesthaker commented on changelog.md:
```diff
@@ -0,0 +1,11 @@
+# Changelog
```
are we doing a manual changelog? github autogenerates one which we can use, esp bc we're not really tracking releases (each PR shouldn't be its own release)

samjgorman replied:
I would _like_ to do a manual changelog to get experience writing a changelog & provide a more concrete ledger of our progress over time.  Perhaps we should remove / modify the concept of releases in the changelog and simplify keep a txt file that we update each week with the major changes we shipped?

samjgorman replied:
I removed this txt file in favor of a Notion page I'll spin up!

kanyesthaker commented on changelog.md:
```diff
@@ -0,0 +1,11 @@
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
+
+## [1.0.0] - 2022-04-16
```
should be v0.1.0 i believe

samjgorman replied:
Modified!

kanyesthaker commented on src/Scene/Boundaries.tsx:
```diff
@@ -0,0 +1,63 @@
+import React from 'react';
+import { PlaneProps } from '../allTypes';
+import Box from '../Shapes/Box';
```
hmmm, I appreciate the reuse of code here to avoid re-implementation, but using 4 boxes as the world boundaries feels a little clunky

samjgorman replied:
Agreed! Resolving this after implementing a `plane` based approach

kanyesthaker commented on src/Scene/Boundaries.tsx:
```diff
@@ -0,0 +1,63 @@
+import Box from '../Shapes/Box';
+
+/**
+ * Boundaries functional component renders four boundaries of a plane given
```
Let's make this a little more functionally descriptive -- explain what the world boundary is vs. exactly how it's implemented

kanyesthaker replied:
for example:

```
Creates a world boundary (4 impassible planes) that define the edges of the world...
```

samjgorman replied:
Great idea! Modified.

kanyesthaker commented on src/Scene/Boundaries.tsx:
```diff
@@ -0,0 +1,63 @@
+  let boundaryHeight;
+
+  if (!props.boundaryHeight) {
+    boundaryHeight = 50;
```
is there any reason for this to not be a constant? why would we need this to be a prop?

samjgorman replied:
This is no longer necessary! Happy to reduce complexity here.

kanyesthaker commented on src/Scene/Boundaries.tsx:
```diff
@@ -0,0 +1,63 @@
+  }
+  return (
+    <>
+      <Box
```
Instead of using 4 boxes, why don't we either
a) use 4 planes to define the world boundary
b) use 1 _really big_ box (collision should work whether you're on the outside or the inside of a shape) (the skybox is effectively option (b))

samjgorman replied:
Okay, I went with option A after doing more research.  I haven't seen a lot of successful examples of what we're attempting to do online, so happy to look through any resources you come across, but I agree that using 4 planes rather than boxes is more robust & eliminates the need to specify a `boundaryHeight`.

samjgorman commented:
Ok @kanyesthaker @shalins , should be ready for re-review. Here's a summary of changes:

1. Introduced a `Plane` component to the shapes library.
2. Refactored `Boundary.tsx` to use 4 impassable positioned & rotated `Plane` components.  This is better than using the `Box` shape because planes can extend infinitely in 2D space to act as impassable boundaries, and removes the need to specify a height. 
3. Refactored `Floor.tsx` to use this `Plane` component.
4. Some QoL updates such as introducing default values for the `type` prop to cut down on the number of required props necessary to render components in the shapes library.

samjgorman commented on src/Player/Player.tsx:
```diff
@@ -37,15 +37,15 @@ export default function Player(props: SphereProps) {
       currentVelocityVector.current.x = round(
         playerVelocity[0],
-        ROUNDING_PRECISION,
+        ROUNDING_PRECISION
```
Note: Looks like a linting issue here. I think this will be resolved in future PRs that use your prettier config, the updated eslintc.json, and a pre-commit hook.

samjgorman commented on src/Scene/Scene.tsx:
```diff
@@ -106,15 +109,13 @@ export default function Scene() {
           />
 
-          <Floor />
-          <gridHelper args={[100, 100, 'black', 'grey']} />
```
Note: Will need to go back and make sure this code uses the constants you've set up in your PR once we've gone and merged

samjgorman replied:
Constants now used after merging the constants PR and resolving merge conflicts

samjgorman commented on src/Shapes/Plane.tsx:
```diff
@@ -0,0 +1,64 @@
+ * @param props customPlane
+ * @returns
+ */
+export default function Plane({ type = 'Static', ...props }: CustomPlane) {
```
One more note: I also didn't modify the structure of the components in the shapes library in this PR, but looking forward to explore approaches to simplify & extend these components in a future PR

samjgorman commented on src/Shapes/Box.tsx:
```diff
@@ -17,34 +18,45 @@ interface CustomBox extends BoxProps {
+        <boxBufferGeometry args={props.dimensions} />
+        <meshPhongMaterial
+          color={props.color}
+          transparent={transparent}
```
Great idea! Similar theme here to video UI PR, great lessons going forward

samjgorman commented on src/Shapes/Box.tsx:
```diff
@@ -28,23 +38,25 @@ export default function Box(props: CustomBox) {
-          <meshPhongMaterial color={props.color} />
-        </mesh>
-      )}
+      <mesh {...meshProps}>
```
I've refactored how meshProps is constructed and hope that this provides more clarity! Now, meshProps looks like this (implemented across the shapes lib)

```typescript
const meshProps = {
    ref: props.collision ? collisionRef : undefined,
    position: props.position,
  };
```

samjgorman commented on src/Scene/Scene.tsx:
```diff
@@ -106,15 +109,13 @@ export default function Scene() {
 
-          <Floor />
-          <gridHelper args={[100, 100, 'black', 'grey']} />
+          <Floor lengthX={PLANE_LENGTH_X} widthZ={PLANE_WIDTH_Z} />
```
Thanks for the feedback!  We now have a `WorldBox` component that functions as a simple wrapper over the `Floor` & `Boundary` components. 

I agree that they should not exist without each other, but would like to keep each component as simple or "pure" as possible.  For example, the `Floor` component may grow in complexity as we introduce new textures or varied terrain, which may become less readable if combined with the `Boundaries` component. 

And in this approach, we still gain the benefits of passing shared props such as `debug` 

Would love your opinion on...
1. Does `WorldBox` make sense for the name of this component?  I see this component as a box extending across the horizontal plane & each four edges.

samjgorman commented:
Ready for rereview @kanyesthaker!

kanyesthaker reviewed (approved):
small nit, otw lgtm

samjgorman commented on src/Scene/WorldBox.tsx:
```diff
@@ -0,0 +1,23 @@
+ * @param props
+ * @returns WorldBox component
+ */
+const WorldBox = (props: WorldBoxProps) => {
```
I think world boundaries is more accurate-- updated! While I was there I updated how we handled props in our shapes library and we now destructure props before using them, which is better style :)
